### ISSUE TITLE
panel-list- and clock-applet icons seem to be wrong size and strange overlap issue with HIDPI resolution, if build out-of-process

### ISSUE THREAD

**Issue by tonyabc** [mate-desktop/mate-panel#818]
#### Expected behaviour
I would expect the Active Apps in the panel to show the icon for that App and be size appropriate for the panel.

#### Actual behaviour
Active App Icons are vertically smaller then they should be and they are not taking the correct space on the panel, also there seems to be 'ghost' icons that are the correct size but not clickable to select the app.

https://imgur.com/a/8gYshw9

#### Steps to reproduce the behaviour
I upgraded from F26 Mate to F28 Mate using DNF and 'proper' F26 --> F27 --> F28 flow using dnf upgrade network option.  Since the new setup the panel has become borked.

#### MATE general version
1.20.2-1,FC28.x86_64

#### Linux Distribution
Fedora 28 Mate Compiz

**Comment by tonyabc** [comment]
Can't work out how to embed so just posted the link

**Comment by raveit65** [comment]
Now the link works.
Are you using HIDPI Monitor? 
This looks like https://github.com/mate-desktop/mate-panel/pull/755#issuecomment-374868245
But i get this problem only sometimes with HIDPI resolutions and switching the workspace fixes the redraw issue.
Btw. Is this report from you?
https://bugzilla.redhat.com/show_bug.cgi?id=1590569#c13

**Comment by tonyabc** [comment]
It's a 4K Screen on a Dell 9550 laptop so yes HIDPI.

Switching workspace doesn't fix the problem, although looking at that I notice the 'bunching' seems only to be firefox active app icons.

Yes I did post that redhat bugzilla as that was a separate bork that happened.

**Comment by raveit65** [comment]
Is this screenshot with compiz?
Compiz don't use workspaces (from technical point).
Can you switch to marco and test if switching workspaces helps?

**Comment by tonyabc** [comment]
A very good point.......slipped my mind I've been using Compiz so long.....

Anyhoo, switching to Marco and switching workspaces didn't fix the issue...

Here's two shots of the Panel, one with a white background and then with black so you can see what's visible.

https://imgur.com/a/fzsM646

**Comment by tonyabc** [comment]
Just to update you and I'm not sure if this is linked or a totally separate bug.

I just opened Libreoffice Writer and it kills the Emerald Theme Manager, even doing a reload of Compiz won't work while Writer is still open.

Until F28 I'd never had a problem with Libroffice and Compiz/Emerald so surprising to get two bugs that seem to be in associated areas.

**Comment by raveit65** [comment]
So it looks here sometimes
![windows-list](https://user-images.githubusercontent.com/961604/42700378-dd731c2c-86c3-11e8-97c9-fd5f46d30446.jpg)

**Comment by lukefromdc** [comment]
Still getting that half-height button bug? That was supposed to have been squashed in one of my commits. I assume that's with wncklet built out-of-process as before?

**Comment by tonyabc** [comment]
Still happening here.....looking at the version though it looks like the Fedora packager hasn't updated from 1.20.2-1,FC28.x86_64 for the Mate components yet.

**Comment by raveit65** [comment]
@lukefromdc 
It still exists if build with out-of-process, i run min. 8 times a day in this issue.
But in my case switching workspaces and back to workspace where the error was, fixes the problem.
In other words i trigger a redraw.
@tonyabc 
If you are a bit familar with rebuilding a SRPM you can build mate-panel with that config option
`--with-in-process-applets=all`
This will fix your problem.
I switched back to `--with-in-process-applets=none` to fix 2 reports where user reports a mem leak with na-tray-applet.
Using 'none' is the traditional way to build the panel, but it causes the redraw issue with HIDPI.

**Comment by lukefromdc** [comment]
It is with and only with wncklet built out-of-process that I've seen this in the past, are you saying you now get it with both out-of-process and with in-process builds of wncklet?

**Comment by raveit65** [comment]
No, i get this only with built out-of-process.
But i can't switch fedora to built in-process because of known issues with na-tray-applet (mem-leaks)

**Comment by lukefromdc** [comment]
Best fix for this might be to cork up that memory leak for in-process, then make this applet in-process only as we've had unending problems with HiDPI when out of process. I don't get a noticable increase in RAM consumption over time though, maybe something in my setup stops the memleak or maybe I just have too much RAM to notice it, so this might have to be done by going through the code looking for places where memory is allocated but not freed.

By any chance is the memory leak only with HiDPI?

**Comment by raveit65** [comment]
No, please see at reports 
We had already a discussion there.
https://github.com/mate-desktop/mate-panel/issues/809
https://bugzilla.redhat.com/show_bug.cgi?id=1576930
https://bugzilla.redhat.com/show_bug.cgi?id=1575091

**Comment by raveit65** [comment]
@lukefromdc 
To make it it clear.
The memleak of na-tray applets or natray itself (build in process), has nothing to do with HIDPI.
It is only the reason why i can't build the panel in-process.
There are people who use fedora-mate on a server 24-7 without shutting down the system.
If you shutdown the sytem every day you may not see the leak.

**Comment by lukefromdc** [comment]
One possible workaround would be to built the tray out of process but the wncklet in-process, thus avoiding each applet's known issues discussed here.

**Comment by raveit65** [comment]
> One possible workaround would be to built the tray out of process but the wncklet in-process, thus avoiding each applet's known issues discussed here.

@lukefromdc @vkareh 
I would be happy with an option to build the wncklist-applet build-in-process independent from all others as workaround. I don't need a commit for it, i would simply use this as patch for fedora :)

**Comment by raveit65** [comment]
Sorry guys, there is already an option to build `wncklet applets` in-process without touching the others.
I will test this.
```
AC_ARG_WITH(in-process-applets,
	AC_HELP_STRING([--with-in-process-applets=APPLETS],
		[comma-separated list of applets to compile in-process (possible values: none, clock, fish, notification-area, wncklet, all) @<:@default=none@:>@]),
```

**Comment by raveit65** [comment]
Btw.  sometimes (very seldom) i noticed the same problem with clock applet if build out-of-process.
in qemu-kvm VM with virtual-viewer.
![clock-applet-kaputt](https://user-images.githubusercontent.com/961604/42939005-8b3617c6-8b54-11e8-9bd9-5c17a90d64c5.png)

**Comment by lukefromdc** [comment]
I wonder if some of this is from GtkPlug/GtkSocket problems? I've noticed bugs in the tray and sometimes also with out-of-process applets with corrupted or wrong icons, and it seems the wrong window contents sometimes gets passed. This looks like the plug contents are not being scaled before being sent to the socket(plug/socket used for all out of process applets) and then are being rendered at both sizes. If the scaling is being done at the plug end, this should not occur unless the plug contents are realized before being scaled.

**Comment by raveit65** [comment]
Why nobody test my rpms which fixes the problem at my box?
I don't push this unconfirmed to fedora repos.
Anyway, I don't see the issue any more ........ i can wait .......... :)

**Comment by samuelgruetter** [comment]
I'm actually on vacation, with limited wifi and limited time for hacking :wink:

However, I think it's awesome that the devs and maintainers of my favorite desktop are so responsive :smiley: 

So I did the following: I went to https://koji.fedoraproject.org/koji/taskinfo?taskID=28404084, clicked on my arch (`x86_64`), which took me to https://koji.fedoraproject.org/koji/taskinfo?taskID=28404085, and from there, I downloaded https://kojipkgs.fedoraproject.org//work/tasks/4085/28404085/mate-panel-1.20.2-2.fc28.x86_64.rpm and https://kojipkgs.fedoraproject.org//work/tasks/4085/28404085/mate-panel-libs-1.20.2-2.fc28.x86_64.rpm.
Then I did

```
sudo dnf install ~/Downloads/mate-panel-1.20.2-2.fc28.x86_64.rpm ~/Downloads/mate-panel-libs-1.20.2-2.fc28.x86_64.rpm 
```

and rebooted.

Now, after opening tons of mate-terminals, my panel looks the way it should:

![no-garbling-but-bad-date](https://user-images.githubusercontent.com/2394355/43037563-b2c66140-8d0e-11e8-9855-3baa0a6c7c98.png)

So the issue I reported at https://github.com/mate-desktop/mate-panel/issues/841 is resolved with the builds posted above by @raveit65 :tada: 

As a minor note, you can see that the clock applet is still garbled, and it seems that this is always the case on my system (but it already happened before), so the builds posted above by @raveit65 are certainly an improvement as far as I can tell from my user's perspective.

**Comment by raveit65** [comment]
Thanks for testing and happy holiday.
I noticed the problem with clock applet already, but i waited that someone confirmed it :smiley: 
Ok, i will do another scratch build with `--with-in-process-applets=wncklet,clock`
Hope the panel starts not leaking memory with having this 2 applets in process.

**Comment by vkareh** [comment]
@raveit65 - good move. If these leak memory they should be easier to fix, as they don't depend on external apps adding icons to them (well, wnck sort of does, but not in the same way as na-tray)

**Comment by samuelgruetter** [comment]
So I just rebooted my machine a few times, and there are many combinations of how correct the right part of my panel is...

Everything ok: 3x

![all-good](https://user-images.githubusercontent.com/2394355/43038211-8f1f8b7a-8d15-11e8-886c-8753543df165.png)

Only the date garbled: 3x

![date-bad](https://user-images.githubusercontent.com/2394355/43038214-9b626470-8d15-11e8-881d-703ca94d660a.png)

Like this: 1x

![date-and-indicator-bad](https://user-images.githubusercontent.com/2394355/43038225-dbdb7384-8d15-11e8-8a8c-8412abf20257.png)

Like this: 1x

![power-bad](https://user-images.githubusercontent.com/2394355/43038234-fa138e2c-8d15-11e8-8b7e-4bff96cc813d.png)

Like this: 1x

![language-switcher-bad](https://user-images.githubusercontent.com/2394355/43038237-063ad5c0-8d16-11e8-9612-68e5eb54c325.png)

Like this: 1x

![bluetooth-and-language-bad](https://user-images.githubusercontent.com/2394355/43038241-152d28e4-8d16-11e8-8f00-8c0cfe7c5de6.png)

Like this: 1x

![bluetooth-lang-power-bad](https://user-images.githubusercontent.com/2394355/43038244-1eddfbde-8d16-11e8-977c-eb2d4d9b0b49.png)

Like this: 1x

![more-garbled](https://user-images.githubusercontent.com/2394355/43038245-26904b84-8d16-11e8-9029-7892ba320e8f.png)

Like this: 1x

![more-bad-2](https://user-images.githubusercontent.com/2394355/43038247-2fe574b6-8d16-11e8-9b78-fb6ccb2a03cd.png)

**Comment by raveit65** [comment]
The icons are all in na-tray-applet. But building the na-tray-applet in-process gives us memory leaks with other external tray-applets.
Beginning with f28 i used `--with-in-process-applets=all` for building the panel.
The result was that users blame me for memory leaks with external tray-applets inside the na-tray-applet.
We have user who use fedora-mate as server which run 24-7.
Here memory leaks are a huge problem.
In other words, i can't do the same **workaround** for na-tray-applet!

Honestly, we need to find out the real cause of this bug if applets are build out-of-process.

**Comment by raveit65** [comment]
@samuelgruetter 
Btw. a brave `killall mate-panel` is sufficient to reload the panel :smile: 
No need restart system or session.

**Comment by tonyabc** [comment]
Sorry, been away so only catching up with things now.

This seems to have fixed my problem.  Will advise if any further issues arise.

Thanks for the prompt fix, appreciated.

**Comment by raveit65** [comment]
Here is the promised build with `--with-in-process-applets=wncklet,clock`
This fixes the problem with clock applet too.
https://koji.fedoraproject.org/koji/taskinfo?taskID=28515377

**Comment by raveit65** [comment]
@vkareh 
I noticed this problem sometimes (after session start) with other panel applets too, eg. netspeed-applet and mate-applet-streamer from https://github.com/assen-totin/mate-applet-streamer 
Rpms are at rpmfusion repo for fedora.
In f28 qemu-kvm VM
netspeed:
![netspeed-applet-bad](https://user-images.githubusercontent.com/961604/43043757-27fe0586-8d9a-11e8-940d-801b20266045.jpg)
Note, you need select the LO device for having zero net-load to avoid a redraw of the applet.
mate-applet-streamer-normal:
![mate-streamer-applet-good](https://user-images.githubusercontent.com/961604/43043763-3d03a300-8d9a-11e8-8f1a-0358ddd5739e.jpg)
mate-applet-streamer-bad:
![mate-streamer-applet-bad](https://user-images.githubusercontent.com/961604/43043766-4ea0b45e-8d9a-11e8-96bd-7264086b59f0.jpg)

I assume that it is more a general problem with the panel because nearly all applets can be affected if build out-of process.
Maybe code from libmate-panel-applet/ folder ?

**Comment by lukefromdc** [comment]
I have been wondering if this is really the same problem as tray icons being corrupted/wrong icon/cut and relates to an issue with GtkPlug/GtkSocket in HiDPI mode.

**Comment by raveit65** [comment]
@samuelgruetter @tonyabc 
A new build for testing changes from merged PR, This is a pre-released version and should fix your issue.
Let me know if that works for you.
f28:
https://koji.fedoraproject.org/koji/taskinfo?taskID=28743847
f27:
https://koji.fedoraproject.org/koji/taskinfo?taskID=28743904


But the real culprit is inside gtk+ itself and is fixed here.
https://gitlab.gnome.org/GNOME/gtk/commit/c8b0a0dea5cd129d8e023443dde5e44319d4631d
This is for gtk+-3.24 which wich release in september.
I had a talk with M.Clasen from gtk+, they will push this new gtk+ release to stable fedora releases.
See https://blog.gtk.org/2018/06/23/a-gtk-3-update/

**Comment by samuelgruetter** [comment]
So I installed `mate-panel-1.20.3-1.fc28.x86_64` and `mate-panel-libs-1.20.3-1.fc28.x86_64` linked above, rebooted, and did a few `killall mate-panel`, but there's no improvement for me, the behavior is still similar to [what I posted before](https://github.com/mate-desktop/mate-panel/issues/818#issuecomment-406808988).
But it's nice to see that at least some other bugs get fixed in the process :wink:

**Comment by raveit65** [comment]
Ok, time is running out....
I will build wncklet and clock applets in-process to fix this annoying issue until gtk+-3.24.x is pushed to fedora stable branches.
Local i have patched latest gtk+-2.22.30 build from fedora with mentioned upstream commit
https://gitlab.gnome.org/GNOME/gtk/commit/c8b0a0dea5cd129d8e023443dde5e44319d4631d
and all the pain is gone :)
If you're familiar with rpm building just do the same.

**Comment by raveit65** [comment]
Guys , i've just submitted an update for f27/28.
Wncklet and clock applets are build in-process.
https://bodhi.fedoraproject.org/updates/FEDORA-2018-b8f25d74ad
https://bodhi.fedoraproject.org/updates/FEDORA-2018-398fd67e4d
This should fix your issue until gtk+ will come with the real fix in september.

**Comment by raveit65** [comment]
I am closing this as fedora builds will be pushed to stable and there is nothing more what we can do.
If you run into this error in other distros file out a report there and ask maintainer to build the panel with wncklet and clock applets in-process.
Any way, real fix will come with gtk+-3.24.0/3.23.1
